base/sparse/spqr.jl

```diff
@@ -21,133 +21,293 @@ const ORDERING_BESTAMD = Int32(9) # try COLAMD and AMD; pick best#
+    if rnk < 0
+        error("Sparse QR factorization failed")
+    end
 
-import Base: size
-import Base.LinAlg: qrfact
-import ..SparseArrays.CHOLMOD: convert, free!
+    e = E[]
+    if e == C_NULL
+        _E = Vector{CHOLMOD.SuiteSparse_long}()
+    else
+        _E = Vector{CHOLMOD.SuiteSparse_long}(n)
+        for i in 1:n
+            @inbounds _E[i] = unsafe_load(e, i) + 1
+        end
+        # Free memory allocated by SPQR. This call will make sure that the
+        # correct deallocator function is called and that the memory count in
+        # the common struct is updated
+        ccall((:cholmod_l_free, :libcholmod), Void,
+            (Csize_t, Cint, Ptr{CHOLMOD.SuiteSparse_long}, Ptr{Void}),
+            n, sizeof(CHOLMOD.SuiteSparse_long), e, CHOLMOD.common())
+    end
+    hpinv = HPinv[]
+    if hpinv == C_NULL
+        _HPinv = Vector{CHOLMOD.SuiteSparse_long}()
+    else
+        _HPinv = Vector{CHOLMOD.SuiteSparse_long}(m)
+        for i in 1:m
+            @inbounds _HPinv[i] = unsafe_load(hpinv, i) + 1
+        end
+        # Free memory allocated by SPQR. This call will make sure that the
+        # correct deallocator function is called and that the memory count in
+        # the common struct is updated
+        ccall((:cholmod_l_free, :libcholmod), Void,
+            (Csize_t, Cint, Ptr{CHOLMOD.SuiteSparse_long}, Ptr{Void}),
+            m, sizeof(CHOLMOD.SuiteSparse_long), hpinv, CHOLMOD.common())
+    end
 
-struct C_Factorization{Tv<:VTypes} <: SuiteSparseStruct
-    xtype::Cint
-    factors::Ptr{Tv}
+    return rnk, _E, _HPinv
 end
 
-mutable struct Factorization{Tv<:VTypes} <: Base.LinAlg.Factorization{Tv}
-    m::Int
-    n::Int
-    p::Ptr{C_Factorization{Tv}}
-    function Factorization{Tv}(m::Integer, n::Integer, p::Ptr{C_Factorization{Tv}}) where Tv<:VTypes
-        if p == C_NULL
-            throw(ArgumentError("factorization failed for unknown reasons. Please submit a bug report."))
-        end
-        new(m, n, p)
-    end
+# Such that A[invperm(p), q] = (I - H[:,1]*τ[1]*H[:,1]')*...*(I - H[:,k]*τ[k]*H[:,k]')*R
+# with k=size(H,2).
+struct QRSparse{Tv,Ti} <: LinAlg.Factorization{Tv}
+    factors::SparseMatrixCSC{Tv,Ti}
+    τ::Vector{Tv}
+    R::SparseMatrixCSC{Tv,Ti}
+    cpiv::Vector{Ti}
+    rpivinv::Vector{Ti}
 end
-Factorization(m::Integer, n::Integer, p::Ptr{C_Factorization{Tv}}) where Tv<:VTypes = Factorization{Tv}(m, n, p)
 
-size(F::Factorization) = (F.m, F.n)
-function size(F::Factorization, i::Integer)
-    if i < 1
-        throw(ArgumentError("dimension must be positive"))
-    end
+Base.size(F::QRSparse) = (size(F.factors, 1), size(F.R, 2))
+function Base.size(F::QRSparse, i::Integer)
     if i == 1
-        return F.m
+        return size(F.factors, 1)
     elseif i == 2
-        return F.n
+        return size(F.R, 2)
+    elseif i > 2
+        return 1
+    else
+        throw(ArgumentError("second argument must be 1 or 2"))
```

Comment:
error seems inconsistent with allowed inputs, any strictly positive integer is accepted